crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -464,8 +481,57 @@ def save_error_state(message = "")
         Rails.logger.error(message)
       end
 
-      def upgrade_compute_nodes
-        # TODO: not implemented
+      def upgrade_all_compute_nodes
+        ["kvm", "xen"].each do |virt|
+          return false unless upgrade_compute_nodes virt
+        end
+        true
+      end
+
+      def upgrade_compute_nodes(virt)
+        save_upgrade_state("Upgrading compute nodes of #{virt} type")
+        compute_nodes = NodeObject.find("roles:nova-compute-#{virt}")
+        return true if compute_nodes.empty?
+
+        controller = NodeObject.find("roles:nova-controller").first
+        if controller.nil?
+          save_error_state("No nova controller node was found!")
+          return false
+        end
+
+        # FIXME: action in this block could be executed in parallel for all compute nodes!
+        # should we use chef-client for that?
```

Comment:
Not a quick one unfortunately. But it definitely makes sense to do this step in parallel and utilizing chef also doesn't sound wrong. OTOH you could also do something similar to what ```ServiceObject::apply_role``` does (starting here: https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/app/models/service_object.rb#L1275) and maybe utilize ```run_remote_chef_client``` (Note: appart from its name that function seems to allow calling other command that "chef-client" as well.)